packages/dart/npt_flutter/lib/features/policy/repositories/role_repository.dart

```diff
@@ -0,0 +1,237 @@
+import 'dart:convert';
+import 'package:at_client_mobile/at_client_mobile.dart';
+import 'package:npt_flutter/app.dart';
+import '../models/policy.dart';
+
+class RoleRepository {
+  static const String groupsPolicyNamespace = 'groups.policy.sshnp';
+
+  Future<List<FetchedRole>> fetchRoles() async {
+    final rolesJson = <String>[];
+    final AtClient atClient = AtClientManager.getInstance().atClient;
+    String? currentAtSign = atClient.getCurrentAtSign();
+    if (currentAtSign == null) {
+      App.log(
+        '[ERROR] fetchRoles: Current atSign is null'.loggable,
+      );
+      return [];
+    }
+    if (!currentAtSign.startsWith('@')) {
+      currentAtSign = '@$currentAtSign';
+    }
+    final String regex = 'groups\\.policy\\.sshnp$currentAtSign';
+
+    final List<String> groupAtKeyStrs = await atClient.getKeys(regex: regex);
+    final List<AtKey> groupAtKeys = groupAtKeyStrs
+        .map((key) => AtKey.fromString(key))
+        .toList();
+
+    for (final AtKey atKey in groupAtKeys) {
+      final GetRequestOptions gro = GetRequestOptions()..useRemoteAtServer = true;
+      AtValue atValue;
+      try {
+        atValue = await atClient.get(atKey, getRequestOptions: gro);
+      } catch (e) {
+        App.log('[ERROR] fetchRoles: Failed to get value for key $atKey: $e ... Continuing anyways :/'
+            .loggable);
+        continue;
+      }
+      if (atValue.value == null) {
+        continue;
+      }
+      final String groupJsonStr = atValue.value;
+      rolesJson.add(groupJsonStr);
+    }
+
+    final roles = <FetchedRole>[];
+
+    for (final roleJsonStr in rolesJson) {
+      try {
+        final roleJson = jsonDecode(roleJsonStr) as Map<String, dynamic>;
+        final role = FetchedRole.fromJson(roleJson);
+        roles.add(role);
+      } catch (e) {
+        App.log('[ERROR] fetchRoles: Failed to parse role JSON: $e'.loggable);
+        continue;
+      }
+    }
+
+    return roles;
+  }
+
+  Future<bool> putNewRole(final RoleInProgress roleInProgress) async {
+    
+    final AtClient atClient = AtClientManager.getInstance().atClient;
+    String? currentAtSign = atClient.getCurrentAtSign();
+
+    if (currentAtSign == null) {
+      App.log(
+        '[ERROR] updateExistingRole: Current atSign is null'.loggable,
+      );
+      return false;
+    }
+
+    // ensure currentAtSign starts with '@'
+    if (!currentAtSign.startsWith('@')) {
+      currentAtSign = '@$currentAtSign';
```

Comment:
tip: you can use .toAtsign() from at_commons, though be warned that toAtsign can throw an Exception if the string contents are invalid. It will add the `@` for you though.